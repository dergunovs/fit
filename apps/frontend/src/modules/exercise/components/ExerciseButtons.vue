<template>
  <div :class="$style.buttons" :data-tall="props.isEdit || props.isPassing">
    <button
      v-if="(props.isEdit || (props.isPassing && !props.isActive && !props.isCurrent)) && props.index !== 0"
      @click="emit('setIndex', props.index)"
      type="button"
      :class="[$style.arrow, $style.pointer, $style.button]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-index-down"
    >
      <IconArrow width="16" height="16" :class="$style.rotateArrow" />
    </button>

    <button
      v-if="(props.isEdit || (props.isPassing && !props.isActive)) && !props.isLast"
      @click="emit('setIndex', props.index + 1)"
      type="button"
      :class="[$style.arrow, $style.pointer, $style.button]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-index-up"
    >
      <IconArrow width="16" height="16" />
    </button>

    <button
      v-if="props.isEdit && props.isSetCreatable"
      @click="emit('createSet')"
      type="button"
      :class="[$style.button, $style.pointer, $style.success]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-create-set"
    >
      +{{ t('set.super') }}
    </button>

    <button
      @click="handleEditRepeats"
      :class="$style.button"
      type="button"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-repeats"
    >
      x{{ props.repeats }}
    </button>

    <button
      @click="handleEditWeight"
      v-if="props.isEdit ? props.isWeights : !!props.weight"
      :class="$style.button"
      type="button"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-weight"
    >
      <IconWeight width="14" height="14" /> {{ props.weight }} {{ t('kg') }}
    </button>

    <span
      v-if="!props.isEdit && !props.isDone && !props.isFutureActivity && !props.isPassing"
      :class="[$style.button, $style.error]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-not-done"
    >
      <IconFail width="14" height="14" /> {{ t('notDone') }}
    </span>

    <button
      @click="handleToggleIsToFailure"
      v-if="(!props.isEdit && props.isToFailure) || (props.isPassing && props.isActive)"
      type="button"
      :class="[$style.button, $style.isToFailure, props.isToFailure && $style.success]"
      :data-tall="props.isEdit || props.isPassing"
      :data-active="props.isActive"
      data-test="exercise-buttons-to-failure"
    >
      <IconToFailure width="14" height="14" /> {{ t('toFailure') }}
    </button>

    <span
      v-if="!props.isEdit && props.duration"
      :class="$style.button"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-duration"
    >
      <IconDuration width="14" height="14" /> {{ formatDuration(props.duration, locale) }}
    </span>

    <button
      v-if="props.isEdit || (props.isPassing && !props.isActive && !props.isOnlyOne)"
      @click="emit('delete')"
      type="button"
      :class="[$style.button, $style.pointer, $style.error]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-delete"
    >
      {{ t('delete') }}
    </button>
  </div>
</template>

<script setup lang="ts">
import { formatDuration } from 'mhz-helpers';

import IconArrow from '@/common/icons/arrow.svg';
import IconWeight from '@/common/icons/weight.svg';
import IconDuration from '@/common/icons/duration.svg';
import IconToFailure from '@/common/icons/to-failure.svg';
import IconFail from '@/common/icons/fail.svg';

import { useTI18n } from '@/common/composables';

interface IProps {
  repeats: number;
  index: number;
  weight?: number;
  isSetCreatable?: boolean;
  isLast?: boolean;
  isEdit?: boolean;
  isDone?: boolean;
  isPassing?: boolean;
  isToFailure?: boolean;
  duration?: number;
  isFutureActivity?: boolean;
  isWeights?: boolean;
  isActive?: boolean;
  isCurrent?: boolean;
  isOnlyOne?: boolean;
}

interface IEmit {
  delete: [];
  createSet: [];
  setIndex: [index: number];
  setIsToFailure: [isToFailure: boolean];
  editRepeats: [];
  editWeight: [];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useTI18n();

function handleEditRepeats() {
  if (props.isEdit || props.isPassing) emit('editRepeats');
}

function handleEditWeight() {
  if (props.isEdit || props.isPassing) emit('editWeight');
}

function handleToggleIsToFailure() {
  if (props.isPassing) emit('setIsToFailure', !props.isToFailure);
}
</script>

<style module lang="scss">
.buttons {
  display: flex;
  align-items: center;
  height: 24px;
  pointer-events: none;
  border: 1px solid var(--color-gray-dark);
  border-radius: 6px;

  &[data-tall='true'] {
    height: 40px;
    pointer-events: all;
  }
}

.button {
  display: flex;
  gap: 2px;
  align-items: center;
  justify-content: center;
  min-width: 32px;
  height: 24px;
  padding: 0 5px;
  font-size: 0.75rem;
  line-height: 1;
  color: var(--color-black);
  cursor: pointer;
  -webkit-user-select: none;
  user-select: none;
  background: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-touch-callout: none;
  border: none;
  border-left: 1px solid var(--color-gray-dark);

  &:first-child {
    border-left: none;
  }

  &:active {
    background: var(--color-gray-light);
    background-color: var(--color-gray-light);
  }

  :focus {
    outline: none;
  }

  &[data-tall='true'] {
    height: 38px;
    padding: 0 6px;
    font-size: 0.875rem;
  }
}

.arrow {
  font-size: 1rem;
  color: var(--color-primary);
}

.rotateArrow {
  rotate: 180deg;
}

.isToFailure {
  color: var(--color-gray-dark);
}

.pointer {
  cursor: pointer;
}

.success {
  color: var(--color-success-dark);
}

.error {
  color: var(--color-error);
}

:global(.dark) {
  .buttons {
    border-color: var(--color-gray-dark-extra);
  }

  .button {
    color: var(--color-white);
    border-left-color: var(--color-gray-dark-extra);
  }

  .isToFailure {
    color: var(--color-gray);
  }

  .success {
    color: var(--color-success);
  }

  .error {
    color: var(--color-error-light);
  }
}
</style>
