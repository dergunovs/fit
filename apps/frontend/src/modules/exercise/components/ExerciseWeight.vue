<template>
  <UiField :label="`${t('weight')}, ${t('kg')}`">
    <UiSelect
      @update:modelValue="(option) => emit('update:modelValue', Number(option))"
      :modelValue="props.modelValue"
      :options="props.options"
      :lang="locale"
      data-test="exercise-weight"
    />
  </UiField>
</template>

<script setup lang="ts">
import { UiField, UiSelect } from 'mhz-ui';

import { useTI18n } from '@/common/composables';

interface IProps {
  options: number[];
  modelValue?: number;
}

interface IEmit {
  'update:modelValue': [option: number];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useTI18n();
</script>
