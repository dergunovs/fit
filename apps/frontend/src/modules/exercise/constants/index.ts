export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;

export const EXERCISE_STATISTICS_HEADERS = [{ title: 'Упражнение' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const MUSCLE_GROUP_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const EXERCISE_LIST_HEADERS = [{ title: 'Название' }];

export const EXERCISE_DELETED_TITLE = 'Упражнение удалено';
