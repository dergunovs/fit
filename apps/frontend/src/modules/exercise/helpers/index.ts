import {
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';
import { createTempId } from 'mhz-helpers';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

export function getUserEquipmentParams(exercise: IExercise, user?: IUser) {
  const isExerciseHasEquipment = !!exercise.equipment;
  const isExerciseHasEquipmentForWeight = !!exercise.equipmentForWeight?.length;
  const isWeightsRequired = !!exercise.isWeightsRequired;

  const isUserHasEquipment = !!user?.equipments?.some(
    (equipment) => equipment.equipment?._id?.toString() === exercise.equipment?._id?.toString()
  );

  const isUserHasEquipmentForWeight = !!user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some(
      (equipmentForWeight) => equipmentForWeight._id?.toString() === equipment.equipment?._id?.toString()
    )
  );

  return {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  };
}

export function isUserEquipmentMatches(exercise: IExercise, user?: IUser) {
  let result = false;

  const {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  } = getUserEquipmentParams(exercise, user);

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (!isExerciseHasEquipment && isExerciseHasEquipmentForWeight && isUserHasEquipmentForWeight) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (
    isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipment &&
    isUserHasEquipmentForWeight
  ) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}`
    : `${exercise.averageDuration.toFixed(1)}`;
}

export function getAverageRepeatsInSet(repeats?: number, sets?: number) {
  return repeats && sets ? Math.round(repeats / sets) : 0;
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise, index) => {
    if (!exercise.dateUpdated || !exercise.duration || !start) return;

    allSteps.push({
      left:
        index === 0
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio,
      right:
        index === exercises.length - 1 && exercises.length !== 1
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    if (!allSteps[index - 1]?.right) return;

    allSteps.push({ left: allSteps[index - 1].right, right: step.left, type: 'rest' });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscle(exercises: IExercise[], title: string, muscle?: string, user?: IUser) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = isUserEquipmentMatches(exercise, user);

    const titleFilter =
      (exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) ||
        exercise.title_en?.toLowerCase().includes(title.toLocaleLowerCase())) &&
      isEquipmentMatches;
    const muscleFilter = exercise.muscles?.some((group) => group._id === muscle) && isEquipmentMatches;

    return muscle ? muscleFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq._id === equipment.equipment?._id)
  );

  if (!equipments?.length) return [];

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?._id === eq._id))
      .map((exToMap) => exToMap._id);

    const availableEquipment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentId) => equipmentId === eq.equipment?._id)
    );

    const options = getWeightsForUserEquipment(availableEquipment);

    return {
      _id: ex._id,
      title: ex.title,
      title_en: ex.title_en,
      options: ex.isWeightsRequired ? options : [0, ...options],
    };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}

export function addSetToExercises(exercises: IExerciseChoosen[]) {
  const set = exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  return [...exercises, ...set];
}

export function updateExercisesIndex(exercises: IExerciseChoosen[], index: number) {
  if (index <= 0 || index >= exercises.length) return exercises;

  const updatedExercises = [...exercises];
  const exerciseToMove = updatedExercises[index];

  updatedExercises.splice(index, 1);
  updatedExercises.splice(index - 1, 0, exerciseToMove);

  return updatedExercises;
}

export function updateExerciseField<T extends keyof IExerciseDone>(
  exercises: IExerciseChoosen[],
  field: T,
  value: IExerciseDone[T],
  id?: string
): IExerciseChoosen[] {
  return exercises.map((exercise) => (exercise._id === id ? { ...exercise, [field]: value } : exercise));
}

export function generateExerciseChoosen(repeats: number, weight?: number, exercise?: IExercise): IExerciseChoosen {
  return {
    _id: createTempId(),
    exercise: {
      _id: exercise?._id || '',
      title: exercise?.title || '',
      title_en: exercise?.title_en || '',
      muscles: exercise?.muscles || [],
      isWeights: exercise?.isWeights || false,
      isWeightsRequired: exercise?.isWeightsRequired || false,
      equipmentForWeight: exercise?.equipmentForWeight,
    },
    repeats,
    weight,
  };
}
