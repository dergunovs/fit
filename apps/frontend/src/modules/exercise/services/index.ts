import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
