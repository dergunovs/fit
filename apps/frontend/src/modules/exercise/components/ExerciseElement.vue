<template>
  <div :class="$style.exercise" :data-hide="props.isHideTitle">
    <UiFlex column gap="2" align="flex-start">
      <div v-if="!props.isHideTitle" data-test="exercise-choosen-title">
        {{ title }}
      </div>

      <ExerciseButtons
        v-if="props.exercise._id"
        :isEdit="props.isEdit"
        :repeats="props.exercise.repeats"
        :weight="props.exercise.weight"
        :index="props.index"
        :isLast="props.isLast"
        :isSetCreatable="props.isSetCreatable"
        :isDone="props.isDone"
        :isToFailure="props.isToFailure"
        :duration="props.duration"
        :isFutureActivity="props.isFutureActivity"
        @createSet="emit('createSet')"
        @delete="emit('delete', props.exercise._id)"
        @setIndex="(updatedIndex) => emit('setIndex', updatedIndex)"
        data-test="exercise-buttons"
      />
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { IExerciseChoosen } from 'fitness-tracker-contracts';
import { UiFlex } from 'mhz-ui';
import { localeField } from 'mhz-helpers';

import ExerciseButtons from '@/exercise/components/ExerciseButtons.vue';

import { useTI18n } from '@/common/composables';

interface IProps {
  exercise: IExerciseChoosen;
  index: number;
  isEdit?: boolean;
  isSetCreatable?: boolean;
  isLast?: boolean;
  isHideTitle?: boolean;
  isDone?: boolean;
  isToFailure?: boolean;
  duration?: number;
  isFutureActivity?: boolean;
}

interface IEmit {
  delete: [id: string];
  createSet: [];
  setIndex: [index: number];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useTI18n();

const title = computed(() => {
  const isExerciseExists = !!props.exercise.exercise;

  if (!isExerciseExists) return t('exercise.deleted');

  const exerciseTitle = props.exercise.exercise?.[localeField('title', locale.value)];

  if (!props.isEdit) return exerciseTitle;

  return `${props.index + 1}. ${exerciseTitle}`;
});
</script>

<style module lang="scss">
.exercise {
  display: flex;
  gap: 12px;
  justify-content: space-between;
  padding: 6px 10px;
  background-color: var(--color-gray-light-extra);
  border-bottom: 1px solid var(--color-gray);
  border-radius: 10px;

  &[data-hide='true'] {
    margin-top: -14px;
    border-top-left-radius: 0;
    border-top-right-radius: 0;
  }
}
</style>
