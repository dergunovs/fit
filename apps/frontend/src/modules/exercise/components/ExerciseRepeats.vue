<template>
  <UiButtongroup
    @update:modelValue="(option) => emit('update:modelValue', option)"
    :modelValue="props.modelValue"
    :options="repeats"
    :title="t('repeat.many')"
    :isTall="props.isTall"
    isInput
    data-test="exercise-repeats"
  />
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { UiButtongroup } from 'mhz-ui';

import { EXERCISE_REPEATS_OPTIONS } from '@/exercise/constants';
import { useTI18n } from '@/common/composables';

interface IProps {
  modelValue: number;
  options?: number[];
  isTall?: boolean;
  baseRepeat?: number;
}

interface IEmit {
  'update:modelValue': [option: number];
}

const props = withDefaults(defineProps<IProps>(), {
  options: () => EXERCISE_REPEATS_OPTIONS,
  baseRepeat: undefined,
});

const emit = defineEmits<IEmit>();

const { t } = useTI18n();

const repeats = computed<number[]>(() =>
  props.baseRepeat
    ? [props.baseRepeat - 2, props.baseRepeat - 1, props.baseRepeat, props.baseRepeat + 1, props.baseRepeat + 2]
    : props.options
);
</script>
