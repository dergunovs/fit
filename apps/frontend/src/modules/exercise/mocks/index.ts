import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
