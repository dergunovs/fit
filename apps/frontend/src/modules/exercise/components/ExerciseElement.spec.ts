import { nextTick } from 'vue';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';
import { IExercise } from 'fitness-tracker-contracts';

import ExerciseElement from './ExerciseElement.vue';
import ExerciseRepeats from '@/exercise/components/ExerciseRepeats.vue';
import ExerciseWeight from '@/exercise/components/ExerciseWeight.vue';
import ExerciseButtons from '@/exercise/components/ExerciseButtons.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISE_CHOOSEN_FIXTURE } from '@/exercise/fixtures';
import { getAvailableExerciseWeights } from '@/exercise/helpers';
import { USER_FIXTURE } from '@/user/fixtures';
import { spyUseAuthCheck } from '@/auth/mocks';

const title = dataTest('exercise-choosen-title');
const buttons = dataTest('exercise-buttons');
const repeats = dataTest('exercise-repeats');
const weight = dataTest('exercise-weight');

const INDEX = 1;
const IS_SET_CREATABLE = true;
const IS_LAST = false;
const IS_EDIT = true;
const IS_DONE = false;
const IS_TO_FAILURE = false;
const DURATION = 0;
const IS_FUTURE_ACTIVITY = false;

let wrapper: VueWrapper<InstanceType<typeof ExerciseElement>>;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseElement, {
    exercise: EXERCISE_CHOOSEN_FIXTURE,
    index: INDEX,
    isSetCreatable: IS_SET_CREATABLE,
    isLast: IS_LAST,
    isEdit: IS_EDIT,
    isDone: IS_DONE,
    isToFailure: IS_TO_FAILURE,
    duration: DURATION,
    isFutureActivity: IS_FUTURE_ACTIVITY,
  });
});

enableAutoUnmount(afterEach);

describe('ExerciseElement', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseElement)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows and hides title', async () => {
    expect(wrapper.find(title).text()).toBe(`${INDEX + 1}. ${EXERCISE_CHOOSEN_FIXTURE.exercise?.title}`);

    await wrapper.setProps({ isEdit: false });

    expect(wrapper.find(title).text()).toBe(EXERCISE_CHOOSEN_FIXTURE.exercise?.title);

    await wrapper.setProps({ isHideTitle: true });

    expect(wrapper.find(title).exists()).toBe(false);
  });

  it('passes props to buttons', async () => {
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('repeats')).toBe(
      EXERCISE_CHOOSEN_FIXTURE.repeats
    );
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('weight')).toBe(
      EXERCISE_CHOOSEN_FIXTURE.weight
    );
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isWeights')).toBe(
      EXERCISE_CHOOSEN_FIXTURE.exercise?.isWeights
    );

    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('index')).toBe(INDEX);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isLast')).toBe(IS_LAST);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isSetCreatable')).toBe(IS_SET_CREATABLE);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isLast')).toBe(IS_LAST);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isEdit')).toBe(IS_EDIT);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isDone')).toBe(IS_DONE);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isToFailure')).toBe(IS_TO_FAILURE);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('duration')).toBe(DURATION);
    expect(wrapper.findComponent<typeof ExerciseButtons>(buttons).props('isFutureActivity')).toBe(IS_FUTURE_ACTIVITY);
  });

  it('emits events from buttons', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('createSet');
    expect(wrapper.emitted()).not.toHaveProperty('delete');
    expect(wrapper.emitted()).not.toHaveProperty('setIndex');

    wrapper.findComponent<typeof ExerciseButtons>(buttons).vm.$emit('createSet');

    expect(wrapper.emitted('createSet')).toHaveLength(1);
    expect(wrapper.emitted()['createSet'][0]).toStrictEqual([]);

    wrapper.findComponent<typeof ExerciseButtons>(buttons).vm.$emit('delete', EXERCISE_CHOOSEN_FIXTURE._id);

    expect(wrapper.emitted('delete')).toHaveLength(1);
    expect(wrapper.emitted()['delete'][0]).toStrictEqual([EXERCISE_CHOOSEN_FIXTURE._id]);

    wrapper.findComponent<typeof ExerciseButtons>(buttons).vm.$emit('setIndex', INDEX + 1);

    expect(wrapper.emitted('setIndex')).toHaveLength(1);
    expect(wrapper.emitted()['setIndex'][0]).toStrictEqual([INDEX + 1]);
  });

  it('updates repeats', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('setRepeats');

    expect(wrapper.find(repeats).isVisible()).toBe(false);

    wrapper.findComponent<typeof ExerciseButtons>(buttons).vm.$emit('editRepeats');

    await nextTick();

    expect(wrapper.find(repeats).isVisible()).toBe(true);

    expect(wrapper.findComponent<typeof ExerciseRepeats>(repeats).props('modelValue')).toStrictEqual(
      EXERCISE_CHOOSEN_FIXTURE.repeats
    );
    expect(wrapper.findComponent<typeof ExerciseRepeats>(repeats).props('baseRepeat')).toStrictEqual(
      EXERCISE_CHOOSEN_FIXTURE.repeats
    );

    const newRepeats = 2;

    wrapper.findComponent<typeof ExerciseRepeats>(repeats).vm.$emit('update:modelValue', newRepeats);

    await nextTick();

    expect(wrapper.emitted('setRepeats')).toHaveLength(1);
    expect(wrapper.emitted()['setRepeats'][0]).toStrictEqual([newRepeats]);

    expect(wrapper.find(repeats).isVisible()).toBe(false);
  });

  it('updates weight', async () => {
    expect(spyUseAuthCheck).toBeCalledTimes(1);

    expect(wrapper.emitted()).not.toHaveProperty('setWeight');

    expect(wrapper.find(weight).isVisible()).toBe(false);

    wrapper.findComponent<typeof ExerciseButtons>(buttons).vm.$emit('editWeight');

    await nextTick();

    expect(wrapper.find(weight).isVisible()).toBe(true);

    expect(wrapper.findComponent<typeof ExerciseWeight>(weight).props('modelValue')).toStrictEqual(
      EXERCISE_CHOOSEN_FIXTURE.weight
    );
    expect(wrapper.findComponent<typeof ExerciseWeight>(weight).props('options')).toStrictEqual(
      getAvailableExerciseWeights(EXERCISE_CHOOSEN_FIXTURE.exercise as IExercise, USER_FIXTURE)
    );

    const newWeight = 20;

    wrapper.findComponent<typeof ExerciseWeight>(weight).vm.$emit('update:modelValue', newWeight);

    await nextTick();

    expect(wrapper.emitted('setWeight')).toHaveLength(1);
    expect(wrapper.emitted()['setWeight'][0]).toStrictEqual([newWeight]);

    expect(wrapper.find(weight).isVisible()).toBe(false);
  });

  it('hides buttons when no exercise id', async () => {
    expect(wrapper.find(buttons).exists()).toBe(true);

    await wrapper.setProps({ exercise: { _id: undefined, repeats: 0 } });

    expect(wrapper.find(buttons).exists()).toBe(false);
  });
});
