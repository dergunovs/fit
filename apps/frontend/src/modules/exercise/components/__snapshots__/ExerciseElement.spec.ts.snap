// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseElement > matches snapshot 1`] = `
"<div class="_exercise_30bc63" data-hide="false">
  <div column="" gap="2" align="flex-start">
    <div data-test="exercise-choosen-title">2. Отжимание от скамьи</div>
    <exercise-buttons-stub repeats="12" weight="0" index="1" duration="0" issetcreatable="true" islast="false" isedit="true" isdone="false" istofailure="false" isfutureactivity="false" data-test="exercise-buttons"></exercise-buttons-stub>
  </div>
</div>"
`;
